code/modules/modular_computers/file_system/programs/generic/ntsl2_interpreter.dm

```diff
@@ -7,93 +7,121 @@
 	requires_ntnet = TRUE
 	available_on_ntnet = TRUE
 
-	nanomodule_path = /datum/nano_module/program/computer_ntsl2_interpreter
-
 	var/datum/ntsl_program/running
+	var/last_terminal_update = 0
 	color = LIGHT_COLOR_GREEN
 
 /datum/computer_file/program/ntsl2_interpreter/process_tick()
 	if(istype(running))
 		running.cycle(30000)
-	..()
+	. = ..()
 
 /datum/computer_file/program/ntsl2_interpreter/kill_program()
-	..()
+	. = ..()
 	if(istype(running))
 		running.kill()
 		running = null
 
+/datum/computer_file/program/ntsl2_interpreter/ui_interact(mob/user)
+	var/datum/vueui/ui = SSvueui.get_open_ui(user, src)
+	if (!ui)
+		ui = new /datum/vueui/modularcomputer(user, src, running ? "mcomputer-generic-ntsl-terminal" : "mcomputer-generic-ntsl-selectfile", 450, 520, "NTSL2+ Interpreter")
+		ui.auto_update_content = TRUE
+	ui.open()
+
+/datum/computer_file/program/ntsl2_interpreter/vueui_transfer(oldobj)
+	for(var/o in SSvueui.transfer_uis(oldobj, src, running ? "mcomputer-generic-ntsl-terminal" : "mcomputer-generic-ntsl-selectfile", 450, 520, "NTSL2+ Interpreter"))
+		var/datum/vueui/ui = o
+		ui.auto_update_content = TRUE
+	return TRUE
+
+/datum/computer_file/program/ntsl2_interpreter/vueui_data_change(var/list/data, var/mob/user, var/datum/vueui/ui)
+	. = ..()
+	data = . || data || list()
+	// Gather data for computer header
+	var/headerdata = get_header_data(data["_PC"])
+	if(headerdata)
+		data["_PC"] = headerdata
+		. = data
+	
+	if(running)
+		if(ui.activeui != "mcomputer-generic-ntsl-terminal")
+			ui.activeui = "mcomputer-generic-ntsl-terminal"
+			spawn(1)
+				. = data
+				ui_interact(user)
+			return .
+		else
```

Comment:
This was done because I was having... issues with switching UIs. I found just reloading the UI after they first moved to a terminal fixed it.

Reply:
It will have issues because you do not fill out data and UI hates when values are being used without them being defined. Also UI will have hard time detecting changes to data if you add new keys to root data list